pennylane/tensorbox/tensorbox.py

```diff
@@ -0,0 +1,353 @@
+
+    __radd__ = __add__
+    __rsub__ = __sub__
+    __rmul__ = __mul__
+
+    @staticmethod
+    def unbox_list(tensors):
+        """Unboxes or unwraps a list of tensor like objects, converting any :class:`TensorBox`
+        objects in the list into raw interface tensors.
+
+        Args:
+            tensors (list[tensor_like]): list of arrays, tensors, or :class:`~.TensorBox` objects
+
+        Returns
+            list[tensor_like]: the input list with all :class:`TensorBox` objects
+            unwrapped
+
+        **Example**
+
+        >>> x = tf.Variable([0.4, 0.1, 0.5])
+        >>> y = TensorBox(x)
+        >>> z = tf.constant([0.1, 0.2])
+
+        Note that this is a static method, so we must pass the tensor represented by the ``TensorBox``
+        if we would like it to be included.
+
+        >>> res = y.unwrap([y, z])
+        >>> res
+        [<tf.Variable 'Variable:0' shape=(3,) dtype=float32, numpy=array([0.4, 0.1, 0.5], dtype=float32)>,
+         <tf.Tensor: shape=(2,), dtype=float32, numpy=array([0.1, 0.2], dtype=float32)>]
+        >>> print([type(v) for v in res])
+        [<class 'tensorflow.python.ops.resource_variable_ops.ResourceVariable'>,
+         <class 'tensorflow.python.framework.ops.EagerTensor'>]
+        """
+        return [v.data if isinstance(v, TensorBox) else v for v in tensors]
+
+    def unbox(self):
+        """Unboxes the ``TensorBox`` container, returning the raw interface tensor."""
+        return self.data
+
+    ###############################################################################
+    # Abstract methods and properties
+    ###############################################################################
+
+    @staticmethod
+    @abc.abstractmethod
+    def astensor(tensor):
+        """Converts the input to the native tensor type of the TensorBox.
+
+        Args:
+            tensor (tensor_like): array to convert
+        """
+
+    @abc.abstractmethod
+    def cast(self, dtype):
+        """Cast the dtype of the TensorBox.
+
+        Args:
+            dtype (np.dtype): The NumPy datatype to cast to.
+                If the boxed tensor is not a NumPy array, the equivalent
+                datastype in the target framework is chosen.
+        """
+
+    @abc.abstractmethod
+    def expand_dims(self, axis):
+        """Expand the shape of the tensor.
+
+        Args:
+            axis (int or tuple[int]): the axis or axes where the additional
+                dimensions should be inserted
+        """
+
+    @property
+    @abc.abstractmethod
+    def interface(self):
+        """str, None: The package that the :class:`.TensorBox` class
+        will dispatch to. The returned strings correspond to those used for PennyLane
+        :doc:`interfaces </introduction/interfaces>`."""
+
+    @abc.abstractmethod
+    def numpy(self):
+        """Converts the tensor to a standard, non-differentiable NumPy ndarray or Python scalar if
+        the tensor is 0-dimensional.

```

Comment:
```suggestion
        """Converts the tensor to a standard, non-differentiable NumPy ndarray, or to a Python scalar if
        the tensor is 0-dimensional.
```maybe this makes it clearer?